buffer/src/main/java/io/netty/buffer/AbstractByteBuf.java

```diff
@@ -1263,7 +1263,41 @@ public String toString(int index, int length, Charset charset) {
 
     @Override
     public int indexOf(int fromIndex, int toIndex, byte value) {
-        return ByteBufUtil.indexOf(this, fromIndex, toIndex, value);
+        if (fromIndex <= toIndex) {
+            return firstIndexOf(fromIndex, toIndex, value);
+        } else {
+            return lastIndexOf(fromIndex, toIndex, value);
+        }
+    }
+
+    private int firstIndexOf(int fromIndex, int toIndex, byte value) {
+        fromIndex = Math.max(fromIndex, 0);
+        if (fromIndex >= toIndex || capacity() == 0) {
+            return -1;
+        }
+
+        for (int i = fromIndex; i < toIndex; i ++) {
```

Comment:
Probably uneeded but I remember and old issue with counted loop causing long time to safe point pauses: maybe would be nice to check if the range is too big and use a long based loop in that case

Reply:
I think it is fine as it is.... this would be true for many other places as well